<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>Selecting biologically important variables with PLSKO</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Selecting biologically important variables
with PLSKO</h1>



<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="co">#library(PLSKO)</span></span></code></pre></div>
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>This vignette illustrates the basic usage of the package to select
important variables in omics data. PLSKO is a method that combines
partial least squares (PLS) regression with knockoff filtering to select
important variables in high-dimensional biological data with false
discovery rate (FDR) control. The package provides functions to generate
the knockoff variables, perform the knockoff filtering and aggregate
multiple knockoff results. The package is designed to be user-friendly
and flexible, allowing users to easily apply the PLSKO method to their
own data.</p>
<div id="overview-of-the-knockoff-framework" class="section level3">
<h3>Overview of the knockoff framework</h3>
<p>Here is the outline of the functions corresponding to the workflow of
knockoff framework:</p>
<table style="width:100%;">
<colgroup>
<col width="16%" />
<col width="11%" />
<col width="15%" />
<col width="34%" />
<col width="10%" />
<col width="12%" />
</colgroup>
<thead>
<tr class="header">
<th></th>
<th></th>
<th></th>
<th></th>
<th><strong>Pipeline Funs Provided^</strong></th>
<th></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><strong>Mais Steps of Knockoff framework</strong></td>
<td><strong>Function</strong></td>
<td><strong>Auxiliary Function</strong></td>
<td><strong>Output (<em>Class</em>)</strong></td>
<td><code>plsko_filter()</code>, <code>plsAKO()</code></td>
<td><code>ko_filter()</code>, <code>AKO_with_KO()</code></td>
</tr>
<tr class="even">
<td><strong>Step 1: Knockoff Variable Generation</strong></td>
<td><code>plsko()</code></td>
<td><code>r_criterion()</code> <br>(for <code>ncomp</code>
estimation)</td>
<td>A <code>n x p</code> <em>matrix</em> of knockoff variables</td>
<td>:heavy_check_mark:</td>
<td>(Bring your own <br>knockoff vars)</td>
</tr>
<tr class="odd">
<td><strong>Step 2: Importance Score Calculation
(<code>W</code>)</strong></td>
<td>(import from pkg <code>Knockoff</code>)</td>
<td>-</td>
<td>A <em>vector</em> of <code>p</code> or a <code>n_ko x p</code>
<em>matrix</em></td>
<td>:heavy_check_mark:</td>
<td>:heavy_check_mark:</td>
</tr>
<tr class="even">
<td><strong>Step 3: Knockoff Filtering and Variable
Selection</strong></td>
<td><code>KO_with_W()</code> <br><code>AKO_with_W()</code></td>
<td>-</td>
<td>A list (class <em>“knockoff.result”</em> or <em>“AKO.result”</em>)
with components: <br> <code>statistic</code>, <code>selected</code>,
<code>ako.selected</code></td>
<td>:heavy_check_mark:</td>
<td>:heavy_check_mark:</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="installation" class="section level2">
<h2>Installation</h2>
<p>You can install the development version of the package from GitHub
using the following code:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="co"># install.packages(&quot;devtools&quot;)</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="co">#devtools::install_github(&quot;guannan-yang/PLSKO/PLSKO&quot;, quiet = TRUE, upgrade = &quot;never&quot;)</span></span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a><span class="fu">library</span>(PLSKO)</span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a></span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a><span class="co">#If warnings of installing dependencies appears, please install the dependencies manually by running the following code:</span></span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a><span class="co"># install.packages(c(&quot;knockoff&quot;,&quot;progress&quot;, &quot;parallel&quot;, &quot;doParallel&quot;, &quot;foreach&quot;))</span></span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a><span class="co"># if (!require(&quot;BiocManager&quot;, quietly = TRUE))</span></span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a><span class="co">#     install.packages(&quot;BiocManager&quot;)</span></span>
<span id="cb2-11"><a href="#cb2-11" tabindex="-1"></a><span class="co"># BiocManager::install(&quot;mixOmics&quot;)</span></span></code></pre></div>
</div>
<div id="example-1-simulated-data" class="section level2">
<h2>Example 1: Simulated data</h2>
<p>In this example, we generate a simulated dataset and response
variable and then show how to apply the PLSKO method to select important
variables.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1</span>)</span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>n <span class="ot">=</span> <span class="dv">100</span> <span class="co"># number of samples</span></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>p <span class="ot">=</span> <span class="dv">150</span> <span class="co"># number of variables ( p &gt; n, high-dimensional setting )</span></span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>k <span class="ot">=</span> <span class="dv">25</span> <span class="co"># number of important variables with nonzeros coefficients</span></span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a>a <span class="ot">=</span> <span class="dv">5</span> <span class="co"># coefficient for important variables</span></span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a></span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a><span class="co"># generate the variables from a multivariate normal distribution</span></span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a>mu <span class="ot">=</span> <span class="fu">rep</span>(<span class="dv">0</span>, p)</span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a>rho <span class="ot">=</span> <span class="fl">0.5</span></span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a>Sigma <span class="ot">=</span> <span class="fu">toeplitz</span>(rho<span class="sc">^</span>(<span class="dv">0</span><span class="sc">:</span>(p<span class="dv">-1</span>))) <span class="co"># we define a covariance matrix with an AR(1) structure</span></span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a>X <span class="ot">=</span> MASS<span class="sc">::</span><span class="fu">mvrnorm</span>(n, mu, Sigma)</span>
<span id="cb3-12"><a href="#cb3-12" tabindex="-1"></a></span>
<span id="cb3-13"><a href="#cb3-13" tabindex="-1"></a><span class="co"># generate a continuous response variable from a linear model</span></span>
<span id="cb3-14"><a href="#cb3-14" tabindex="-1"></a>nonzero <span class="ot">=</span> <span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span>p, k)</span>
<span id="cb3-15"><a href="#cb3-15" tabindex="-1"></a>beta <span class="ot">=</span> a <span class="sc">*</span> (<span class="dv">1</span><span class="sc">:</span>p <span class="sc">%in%</span> nonzero) <span class="sc">/</span> <span class="fu">sqrt</span>(n)</span>
<span id="cb3-16"><a href="#cb3-16" tabindex="-1"></a>y <span class="ot">=</span> X <span class="sc">%*%</span> beta <span class="sc">+</span> <span class="fu">rnorm</span>(n)</span>
<span id="cb3-17"><a href="#cb3-17" tabindex="-1"></a></span>
<span id="cb3-18"><a href="#cb3-18" tabindex="-1"></a><span class="co"># generate a binary response variable from binomial with sigmoid function as the probability</span></span>
<span id="cb3-19"><a href="#cb3-19" tabindex="-1"></a>y_bin <span class="ot">=</span> <span class="fu">as.factor</span>(<span class="fu">rbinom</span>(n, <span class="dv">1</span>, <span class="fu">plogis</span>(X <span class="sc">%*%</span> beta)))</span></code></pre></div>
<p>In the generated X matrix, the variables are correlated with an AR(1)
structures (i.e., each entry <span class="math inline">\(\rho_{ij} =
\rho^{|ij|}\)</span>), which means correlations to be higher between
variables that are closer to each other. (And we will use this
information to define the neighbourhoods of variables in the PLSKO
method in the following example.)</p>
<p>We then generate the response variable y from a linear model with 25
important variables and 175 unimportant variables. The important
variables have non-zero coefficients, while the unimportant variables
have zero coefficients.</p>
<p>Next, we apply the PLSKO method to select important variables in the
simulated data.</p>
<div id="plsko-pipeline-apply-knockoff-framework-with-plsko-generated-knockoff-in-a-single-function" class="section level3">
<h3>PLSKO pipeline: apply knockoff framework with PLSKO-generated
knockoff in a single function</h3>
<p>If you are new to knockoff, we can use the function to perform the
knockoff filtering.</p>
<div id="default-settings" class="section level4">
<h4>Default settings</h4>
<p>First let’s run with the default settings. With the default settings,
the function requires the predictor matrix , the response vector as
input arguments. And it returns an object of class . You can print the
result to see the selected variables.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="co"># run the knockoff filter with default settings</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a>result <span class="ot">=</span> <span class="fu">plsko_filter</span>(X, y) </span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a><span class="fu">print</span>(result)</span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a><span class="co">#&gt; Call:</span></span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a><span class="co">#&gt; plsko_filter(X = X, y = y)</span></span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a><span class="co">#&gt; Selected variables:</span></span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a><span class="co">#&gt; [1]   9  23  61  62  73  82 107 146</span></span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a></span>
<span id="cb4-10"><a href="#cb4-10" tabindex="-1"></a><span class="co"># compare with the true coefficients</span></span>
<span id="cb4-11"><a href="#cb4-11" tabindex="-1"></a><span class="fu">which</span>(beta <span class="sc">!=</span> <span class="dv">0</span>)</span>
<span id="cb4-12"><a href="#cb4-12" tabindex="-1"></a><span class="co">#&gt;  [1]   9  20  23  31  44  48  50  53  55  61  62  63  66  71  73  78  82  97 105</span></span>
<span id="cb4-13"><a href="#cb4-13" tabindex="-1"></a><span class="co">#&gt; [20] 107 118 121 125 146 147</span></span>
<span id="cb4-14"><a href="#cb4-14" tabindex="-1"></a></span>
<span id="cb4-15"><a href="#cb4-15" tabindex="-1"></a><span class="co"># calculate FDP in this run</span></span>
<span id="cb4-16"><a href="#cb4-16" tabindex="-1"></a>fdp <span class="ot">&lt;-</span> <span class="cf">function</span>(result) {</span>
<span id="cb4-17"><a href="#cb4-17" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">class</span>(result) <span class="sc">==</span> <span class="st">&quot;knockoff.result&quot;</span>) fdp <span class="ot">=</span> <span class="fu">sum</span>(beta[result<span class="sc">$</span>selected] <span class="sc">==</span><span class="dv">0</span>) <span class="sc">/</span> <span class="fu">max</span>(<span class="fu">length</span>(result<span class="sc">$</span>selected), <span class="dv">1</span>)</span>
<span id="cb4-18"><a href="#cb4-18" tabindex="-1"></a>  <span class="cf">else</span> <span class="cf">if</span> (<span class="fu">class</span>(result) <span class="sc">==</span> <span class="st">&quot;AKO.result&quot;</span>) fdp <span class="ot">=</span> <span class="fu">sum</span>(beta[result<span class="sc">$</span>ako.s] <span class="sc">==</span><span class="dv">0</span>) <span class="sc">/</span> <span class="fu">max</span>(<span class="fu">length</span>(result<span class="sc">$</span>ako.s), <span class="dv">1</span>)</span>
<span id="cb4-19"><a href="#cb4-19" tabindex="-1"></a>  <span class="fu">return</span>(fdp)</span>
<span id="cb4-20"><a href="#cb4-20" tabindex="-1"></a>}</span>
<span id="cb4-21"><a href="#cb4-21" tabindex="-1"></a><span class="fu">fdp</span>(result)</span>
<span id="cb4-22"><a href="#cb4-22" tabindex="-1"></a><span class="co">#&gt; [1] 0</span></span></code></pre></div>
<p>The default settings, which are: neighbourhoods are determined based
on 80-quantile of the sample correlations, the number of components is
determined empirically by the criterion (minimum 2), and the sparsity
level is set to 1 (no sparsity) for PLS regression and the target FDR
level is 0.05. The result shows that the PLSKO method selected the 8
important variables with a false discovery proportion (FDP) of 0.</p>
</div>
<div id="customized-settings-example-1-customised-neighbourhood-information" class="section level4">
<h4>Customized settings example 1: Customised neighbourhood
information</h4>
<p>You can also customize the settings of the PLSKO method by specifying
the parameters in the function. For example, you can specify the
neighbourhood information, the number of components, the sparsity level
in PLS regression, and the target FDR level, etc.</p>
<p>First, we define the neighbourhood information based on the AR(1)
structure of the variables (which we know is true in this case, in real
data, you might need to estimate the neighbourhood information from the
data or make some assumptions based on domain knowledge). Specifically,
we define the neighbourhood of each variable as the variables that are
within a distance of 3 variables from the variable.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="co"># define the neighbourhood information based on the AR(1) structure of the variables</span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a><span class="co"># Option 1: define the neighbourhood as a list of length p</span></span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a>nb.list <span class="ot">=</span> <span class="fu">lapply</span>(<span class="dv">1</span><span class="sc">:</span>p, <span class="cf">function</span>(i){</span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a>  <span class="fu">c</span>((i<span class="dv">-3</span>)<span class="sc">:</span>(i<span class="dv">-1</span>), (i<span class="sc">+</span><span class="dv">1</span>)<span class="sc">:</span>(i<span class="sc">+</span><span class="dv">3</span>))</span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a>})</span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a><span class="co"># remove the indices that are out of the range</span></span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a>nb.list <span class="ot">=</span> <span class="fu">lapply</span>(nb.list, <span class="cf">function</span>(x) x[x <span class="sc">&gt;</span> <span class="dv">0</span> <span class="sc">&amp;</span> x <span class="sc">&lt;=</span> p])</span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a></span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a><span class="co"># Then, we run the PLSKO method with the customized neighbourhood information. </span></span>
<span id="cb5-10"><a href="#cb5-10" tabindex="-1"></a>result <span class="ot">=</span> <span class="fu">plsko_filter</span>(X, y, <span class="at">nb.list =</span> nb.list)</span>
<span id="cb5-11"><a href="#cb5-11" tabindex="-1"></a><span class="fu">print</span>(result)</span>
<span id="cb5-12"><a href="#cb5-12" tabindex="-1"></a><span class="co">#&gt; Call:</span></span>
<span id="cb5-13"><a href="#cb5-13" tabindex="-1"></a><span class="co">#&gt; plsko_filter(X = X, y = y, nb.list = nb.list)</span></span>
<span id="cb5-14"><a href="#cb5-14" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb5-15"><a href="#cb5-15" tabindex="-1"></a><span class="co">#&gt; Selected variables:</span></span>
<span id="cb5-16"><a href="#cb5-16" tabindex="-1"></a><span class="co">#&gt;  [1]   9  23  50  53  54  63  73  82 107 118 121 146</span></span>
<span id="cb5-17"><a href="#cb5-17" tabindex="-1"></a><span class="fu">fdp</span>(result)</span>
<span id="cb5-18"><a href="#cb5-18" tabindex="-1"></a><span class="co">#&gt; [1] 0.08333333</span></span>
<span id="cb5-19"><a href="#cb5-19" tabindex="-1"></a></span>
<span id="cb5-20"><a href="#cb5-20" tabindex="-1"></a><span class="co"># Option 2: define the neighbourhood as an adjacency matrix</span></span>
<span id="cb5-21"><a href="#cb5-21" tabindex="-1"></a>nb.mat <span class="ot">=</span> <span class="fu">matrix</span>(<span class="dv">0</span>, p, p)</span>
<span id="cb5-22"><a href="#cb5-22" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>p){</span>
<span id="cb5-23"><a href="#cb5-23" tabindex="-1"></a>  <span class="co"># make sure the indices are within the range</span></span>
<span id="cb5-24"><a href="#cb5-24" tabindex="-1"></a>  nb <span class="ot">=</span> (i<span class="dv">-3</span>)<span class="sc">:</span>(i<span class="sc">+</span><span class="dv">3</span>)</span>
<span id="cb5-25"><a href="#cb5-25" tabindex="-1"></a>  nb <span class="ot">=</span> nb[nb <span class="sc">&gt;</span> <span class="dv">0</span> <span class="sc">&amp;</span> nb <span class="sc">&lt;=</span> p]</span>
<span id="cb5-26"><a href="#cb5-26" tabindex="-1"></a>  nb.mat[i, nb] <span class="ot">=</span> <span class="dv">1</span></span>
<span id="cb5-27"><a href="#cb5-27" tabindex="-1"></a>}</span>
<span id="cb5-28"><a href="#cb5-28" tabindex="-1"></a><span class="fu">isSymmetric</span>(nb.mat) <span class="co"># check if the matrix is symmetric</span></span>
<span id="cb5-29"><a href="#cb5-29" tabindex="-1"></a><span class="co">#&gt; [1] TRUE</span></span>
<span id="cb5-30"><a href="#cb5-30" tabindex="-1"></a></span>
<span id="cb5-31"><a href="#cb5-31" tabindex="-1"></a>result <span class="ot">=</span> <span class="fu">plsko_filter</span>(X, y, <span class="at">nb.list =</span> nb.mat)</span>
<span id="cb5-32"><a href="#cb5-32" tabindex="-1"></a><span class="co">#&gt; Warning in plsko(X, nb.list = nb.list, threshold.abs = threshold.abs,</span></span>
<span id="cb5-33"><a href="#cb5-33" tabindex="-1"></a><span class="co">#&gt; threshold.q = threshold.q, : Input nb.list must have 0 on the diagonal. Setting</span></span>
<span id="cb5-34"><a href="#cb5-34" tabindex="-1"></a><span class="co">#&gt; diagonal to 0.</span></span>
<span id="cb5-35"><a href="#cb5-35" tabindex="-1"></a><span class="fu">print</span>(result)</span>
<span id="cb5-36"><a href="#cb5-36" tabindex="-1"></a><span class="co">#&gt; Call:</span></span>
<span id="cb5-37"><a href="#cb5-37" tabindex="-1"></a><span class="co">#&gt; plsko_filter(X = X, y = y, nb.list = nb.mat)</span></span>
<span id="cb5-38"><a href="#cb5-38" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb5-39"><a href="#cb5-39" tabindex="-1"></a><span class="co">#&gt; Selected variables:</span></span>
<span id="cb5-40"><a href="#cb5-40" tabindex="-1"></a><span class="co">#&gt;  [1]   9  23  50  53  54  63  73  82 107 118 121 146</span></span>
<span id="cb5-41"><a href="#cb5-41" tabindex="-1"></a><span class="fu">fdp</span>(result)</span>
<span id="cb5-42"><a href="#cb5-42" tabindex="-1"></a><span class="co">#&gt; [1] 0.08333333</span></span></code></pre></div>
<p>These two options are equivalent, and you can choose the one that is
more convenient for you. The result shows that the PLSKO method selected
12 important variables with a FDP of 0.833, which is a little bit higher
than the target FDR level of 0.05. This might due to the small sample
size and the high-dimensional setting, modified FDR control by
<code>offset = 0</code> is used in this run, or purely randomness in
knockoff variable generating (given knockoff frame work only ensure FDR
control as an expectation of FDP, which might fluctuate in different
runs). However, the power of the PLSKO method is better than the default
run with input of neighbourhood information.</p>
</div>
<div id="customized-settings-example-2-customised-number-of-components-and-sparsity-level" class="section level4">
<h4>Customized settings example 2: Customised number of components and
sparsity level</h4>
<p>You can also specify the number of components and the sparsity level
in the PLS regression. For example, you can set the number of components
to 3 and the sparsity level to 0.9.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="co"># run the PLSKO method with the number of components set to 3 and the sparsity level set to 0.9, which means 90% of the coefficients in PLS regression are zero on each component.</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>result <span class="ot">=</span> <span class="fu">plsko_filter</span>(X, y, <span class="at">ncomp =</span> <span class="dv">3</span>, <span class="at">sparsity =</span> <span class="fl">0.95</span>)</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a><span class="fu">print</span>(result)</span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a><span class="co">#&gt; Call:</span></span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a><span class="co">#&gt; plsko_filter(X = X, y = y, ncomp = 3, sparsity = 0.95)</span></span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a><span class="co">#&gt; Selected variables:</span></span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a><span class="co">#&gt; [1]  61  62  82 107 146</span></span>
<span id="cb6-9"><a href="#cb6-9" tabindex="-1"></a><span class="fu">fdp</span>(result)</span>
<span id="cb6-10"><a href="#cb6-10" tabindex="-1"></a><span class="co">#&gt; [1] 0</span></span></code></pre></div>
<p>We observed that the PLSKO method selected 5 important variables (out
of 25) with a FDP of 0.</p>
</div>
<div id="customized-settings-example-3-binary-response" class="section level4">
<h4>Customized settings example 3: Binary response</h4>
<p>You can also apply the PLSKO method to a binary response variable. In
this case, you need to specify the method to compute the test
statistics. For example, you can set the method to “lasso.logistic” to
use the difference of coefficients in LASSO logistic regression. Or
without specifying, <code>plsko_pipeline</code> will automatically
adjust the method based on the response type.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="co"># run the knockoff filter with default settings for binary response</span></span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a>result <span class="ot">=</span> <span class="fu">plsko_filter</span>(X, y_bin)</span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a><span class="co">#&gt; Warning in ko_filter(X, Xk, y, q = q, method = method, offset = offset):</span></span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a><span class="co">#&gt; lasso.lcd is not applicable for categorical response variable, change to</span></span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a><span class="co">#&gt; lasso.logistic</span></span>
<span id="cb7-6"><a href="#cb7-6" tabindex="-1"></a><span class="fu">print</span>(result)</span>
<span id="cb7-7"><a href="#cb7-7" tabindex="-1"></a><span class="co">#&gt; Call:</span></span>
<span id="cb7-8"><a href="#cb7-8" tabindex="-1"></a><span class="co">#&gt; plsko_filter(X = X, y = y_bin)</span></span>
<span id="cb7-9"><a href="#cb7-9" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb7-10"><a href="#cb7-10" tabindex="-1"></a><span class="co">#&gt; Selected variables:</span></span>
<span id="cb7-11"><a href="#cb7-11" tabindex="-1"></a><span class="co">#&gt; NULL</span></span>
<span id="cb7-12"><a href="#cb7-12" tabindex="-1"></a><span class="fu">fdp</span>(result)</span>
<span id="cb7-13"><a href="#cb7-13" tabindex="-1"></a><span class="co">#&gt; [1] 0</span></span></code></pre></div>
<p>We observed <code>NULL</code> in the result, which means no variable
is selected in this run. This might be due to the small sample size. In
binary response, given more randomness in the response, the power of the
PLSKO method is generally lower than in continuous response, and you
might need a larger sample size to achieve good power.</p>
</div>
</div>
<div id="pls-ako-pipeline-aggregate-multiple-knockoff-results-from-plsko-method" class="section level3">
<h3>PLS-AKO pipeline: aggregate multiple knockoff results from
<code>PLSKO</code> method</h3>
<p>Given the randomness in the knockoff variable generation and the PLS
regression, the results of the PLSKO method might vary in different
runs. To improve the stability and power of the variable selection, you
can aggregate multiple knockoff results using the PLS-AKO method. We
only show the default settings here, you can also customize the settings
in the function, such as the number of iterations (<code>n_ko</code>)
and other parameters similar to the function. The default setting is to
run 25 iterations of the PLSKO method and aggregate the results using
the PLS-AKO method. We provide options of parallel computing (default)
in the function, which can significantly reduce the computation time
when running multiple iterations.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="co"># run the PLS-AKO method with default settings</span></span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a>result <span class="ot">=</span> <span class="fu">plsAKO</span>(X, y)</span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a><span class="fu">print</span>(result)</span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a><span class="co">#&gt; Call:</span></span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a><span class="co">#&gt; plsAKO(X = X, y = y)</span></span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a><span class="co">#&gt; Selected variables (AKO):</span></span>
<span id="cb8-8"><a href="#cb8-8" tabindex="-1"></a><span class="co">#&gt; [1]  61  62  73 107 146</span></span>
<span id="cb8-9"><a href="#cb8-9" tabindex="-1"></a><span class="co">#&gt; Frequency of selected variables from 25 iterations of single-run knockoffs:</span></span>
<span id="cb8-10"><a href="#cb8-10" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb8-11"><a href="#cb8-11" tabindex="-1"></a><span class="co">#&gt;   6   9  23  27  31  45  48  50  51  53  55  60  61  62  63  73  75  78  82 107 </span></span>
<span id="cb8-12"><a href="#cb8-12" tabindex="-1"></a><span class="co">#&gt;   1   3   8   1   1   1   1   1   1   2   5   1  14  11   4  11   1   3   7  18 </span></span>
<span id="cb8-13"><a href="#cb8-13" tabindex="-1"></a><span class="co">#&gt; 114 118 119 120 121 125 146 </span></span>
<span id="cb8-14"><a href="#cb8-14" tabindex="-1"></a><span class="co">#&gt;   1   4   7   4   7   1  16</span></span>
<span id="cb8-15"><a href="#cb8-15" tabindex="-1"></a><span class="fu">fdp</span>(result)</span>
<span id="cb8-16"><a href="#cb8-16" tabindex="-1"></a><span class="co">#&gt; [1] 0</span></span>
<span id="cb8-17"><a href="#cb8-17" tabindex="-1"></a></span>
<span id="cb8-18"><a href="#cb8-18" tabindex="-1"></a><span class="co">#Binary response</span></span>
<span id="cb8-19"><a href="#cb8-19" tabindex="-1"></a><span class="co"># result = plsAKO(X, y_bin)</span></span>
<span id="cb8-20"><a href="#cb8-20" tabindex="-1"></a><span class="co"># print(result)</span></span>
<span id="cb8-21"><a href="#cb8-21" tabindex="-1"></a><span class="co"># fdp(result)</span></span></code></pre></div>
<p>The result shows that the PLS-AKO method selected 5 important
variables with a FDP of 0,</p>
</div>
</div>
<div id="example-2-semi-synthetic-data-with-continuous-response" class="section level2">
<h2>Example 2: Semi-synthetic data with continuous response</h2>
<p>In this example, we generate a semi-synthetic dataset with continuous
response and apply the PLSKO method to select important variables in
biological data.</p>
<p>There are two datasets provided in the package:
<code>prot_placenta</code> and <code>cfRNA_placenta</code>. The
<code>prot_placenta</code> dataset contains relative abundances of
proteins from 36 samples with 36 genes. The <code>rna_placenta</code>
dataset contains cell-free RNA counts (Moufarrej et al. ,2022) from 71
samples with 81 genes with elevated expression in placenta. The proteins
are a subset of proteomics data (from a multi-omics pre-eclampsia study
(Marić et al. 2022)) that were inferred released by placenta, according
to Degnes et al. (2022).</p>
<div id="semi-synthetic-data-based-on-the-prot_placenta-dataset" class="section level3">
<h3>Semi-synthetic data based on the <code>prot_placenta</code>
dataset</h3>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="fu">data</span>(<span class="st">&quot;prot_placenta&quot;</span>)</span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a>X <span class="ot">=</span> <span class="fu">as.matrix</span>(prot_placenta<span class="sc">$</span>abundance)</span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a></span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a><span class="co">#generate the response variable y from a linear model</span></span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1</span>)</span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a>n <span class="ot">=</span> <span class="fu">nrow</span>(X)</span>
<span id="cb9-7"><a href="#cb9-7" tabindex="-1"></a>p <span class="ot">=</span> <span class="fu">ncol</span>(X)</span>
<span id="cb9-8"><a href="#cb9-8" tabindex="-1"></a>k <span class="ot">=</span> <span class="dv">8</span> <span class="co"># number of important variables with nonzeros coefficients</span></span>
<span id="cb9-9"><a href="#cb9-9" tabindex="-1"></a>nonzero <span class="ot">=</span> <span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span>p, k) <span class="co"># randomly select 8 important variables</span></span>
<span id="cb9-10"><a href="#cb9-10" tabindex="-1"></a></span>
<span id="cb9-11"><a href="#cb9-11" tabindex="-1"></a>beta <span class="ot">=</span> <span class="fu">as.numeric</span>(<span class="dv">1</span><span class="sc">:</span>p <span class="sc">%in%</span> nonzero) <span class="co"># assign non-zero coefficients to the important variables</span></span>
<span id="cb9-12"><a href="#cb9-12" tabindex="-1"></a>y <span class="ot">=</span> X <span class="sc">%*%</span> beta </span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a>result <span class="ot">=</span> <span class="fu">plsko_filter</span>(X, y) </span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a><span class="fu">print</span>(result)</span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a><span class="co">#&gt; Call:</span></span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a><span class="co">#&gt; plsko_filter(X = X, y = y)</span></span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb10-6"><a href="#cb10-6" tabindex="-1"></a><span class="co">#&gt; Selected variables:</span></span>
<span id="cb10-7"><a href="#cb10-7" tabindex="-1"></a><span class="co">#&gt;    HSPB1    TIMP3 SERPINE1   DIABLO SERPINE2     TFPI  ANGPTL4    GREM1 </span></span>
<span id="cb10-8"><a href="#cb10-8" tabindex="-1"></a><span class="co">#&gt;        1        4       11       14       15       18       23       27 </span></span>
<span id="cb10-9"><a href="#cb10-9" tabindex="-1"></a><span class="co">#&gt;      IL6      NOG </span></span>
<span id="cb10-10"><a href="#cb10-10" tabindex="-1"></a><span class="co">#&gt;       29       34</span></span>
<span id="cb10-11"><a href="#cb10-11" tabindex="-1"></a><span class="fu">fdp</span>(result)</span>
<span id="cb10-12"><a href="#cb10-12" tabindex="-1"></a><span class="co">#&gt; [1] 0.2</span></span></code></pre></div>
<p>The result shows that the PLSKO method selected 10 important
variables with a FDP of 0.2 (all 8 true positive plus 2 false
discoveries), higher than the default FDR level of 0.05. This might be
due to the small number of important variables.</p>
<p>Another reason is that in real data, the true distrution is unknown
and highly correlated, we might want to include more neighbours to
control the correlations among variables, plus include more components
in PLS regression to capture the structure of the data.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a>result <span class="ot">&lt;-</span> <span class="fu">plsko_filter</span>(X, y, <span class="at">threshold.abs =</span> <span class="dv">0</span>, <span class="at">ncomp =</span> <span class="dv">5</span>, <span class="at">sparsity =</span> <span class="fl">0.8</span>) <span class="co"># set the absolute correlation threshold to 0 (every one is neighour with each other), the number of components to 5, and the sparsity level in PLS regression to 0.8</span></span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a><span class="fu">print</span>(result)</span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a><span class="co">#&gt; Call:</span></span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a><span class="co">#&gt; plsko_filter(X = X, y = y, threshold.abs = 0, ncomp = 5, sparsity = 0.8)</span></span>
<span id="cb11-5"><a href="#cb11-5" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb11-6"><a href="#cb11-6" tabindex="-1"></a><span class="co">#&gt; Selected variables:</span></span>
<span id="cb11-7"><a href="#cb11-7" tabindex="-1"></a><span class="co">#&gt;    HSPB1    TIMP3 SERPINE1   DIABLO     TFPI  ANGPTL4    GREM1      NOG </span></span>
<span id="cb11-8"><a href="#cb11-8" tabindex="-1"></a><span class="co">#&gt;        1        4       11       14       18       23       27       34</span></span>
<span id="cb11-9"><a href="#cb11-9" tabindex="-1"></a><span class="fu">fdp</span>(result)</span>
<span id="cb11-10"><a href="#cb11-10" tabindex="-1"></a><span class="co">#&gt; [1] 0</span></span></code></pre></div>
<p>We observed that the PLSKO method selected all 8 important variables
with a FDP of 0.</p>
<p>Let’s see how aggregating multiple knockoff results using the PLS-AKO
method.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a>result <span class="ot">=</span> <span class="fu">plsAKO</span>(X, y, <span class="at">threshold.abs =</span> <span class="dv">0</span>, <span class="at">ncomp =</span> <span class="dv">5</span>, <span class="at">sparsity =</span> <span class="fl">0.8</span>)</span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a><span class="fu">print</span>(result)</span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a><span class="co">#&gt; Call:</span></span>
<span id="cb12-4"><a href="#cb12-4" tabindex="-1"></a><span class="co">#&gt; plsAKO(X = X, y = y, threshold.abs = 0, ncomp = 5, sparsity = 0.8)</span></span>
<span id="cb12-5"><a href="#cb12-5" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb12-6"><a href="#cb12-6" tabindex="-1"></a><span class="co">#&gt; Selected variables (AKO):</span></span>
<span id="cb12-7"><a href="#cb12-7" tabindex="-1"></a><span class="co">#&gt;    HSPB1    TIMP3 SERPINE1   DIABLO     TFPI  ANGPTL4    GREM1      NOG </span></span>
<span id="cb12-8"><a href="#cb12-8" tabindex="-1"></a><span class="co">#&gt;        1        4       11       14       18       23       27       34 </span></span>
<span id="cb12-9"><a href="#cb12-9" tabindex="-1"></a><span class="co">#&gt; Frequency of selected variables from 25 iterations of single-run knockoffs:</span></span>
<span id="cb12-10"><a href="#cb12-10" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb12-11"><a href="#cb12-11" tabindex="-1"></a><span class="co">#&gt;  1  4 11 14 15 18 23 27 29 34 </span></span>
<span id="cb12-12"><a href="#cb12-12" tabindex="-1"></a><span class="co">#&gt; 25 25 25 25  4 25 25 25  4 25</span></span>
<span id="cb12-13"><a href="#cb12-13" tabindex="-1"></a><span class="fu">fdp</span>(result)</span>
<span id="cb12-14"><a href="#cb12-14" tabindex="-1"></a><span class="co">#&gt; [1] 0</span></span></code></pre></div>
<p>Similar to the PLSKO method, the PLS-AKO method selected all 8
important variables with a FDP of 0, although 7 of these 25 iteration
include at one false discovery. This shows how the PLS-AKO method can
improve the stability of the result.</p>
</div>
</div>
<div id="example-3-advanced-usage-of-plsko-package" class="section level2">
<h2>Example 3: Advanced usage of <code>PLSKO</code> package</h2>
<p>Now as you know the basic usage of the <code>PLSKO</code> package, we
will show you how to use the package in a more advanced way. Our package
also provides separate function for each step of the knockoff framework,
which allows users to have more flexibility in the workflow. In this
example, we show how to use the <code>PLSKO</code> package to apply the
PLSKO method to the <code>cfRNA_placenta</code> dataset with a synthetic
response variable.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="fu">data</span>(<span class="st">&quot;cfRNA_placenta&quot;</span>)</span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a>X <span class="ot">=</span> <span class="fu">as.matrix</span>(cfRNA_placenta<span class="sc">$</span>counts)</span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a><span class="co">#generate the response variable y from a linear model</span></span>
<span id="cb13-4"><a href="#cb13-4" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1</span>)</span>
<span id="cb13-5"><a href="#cb13-5" tabindex="-1"></a>n <span class="ot">=</span> <span class="fu">nrow</span>(X)</span>
<span id="cb13-6"><a href="#cb13-6" tabindex="-1"></a>p <span class="ot">=</span> <span class="fu">ncol</span>(X)</span>
<span id="cb13-7"><a href="#cb13-7" tabindex="-1"></a>k <span class="ot">=</span> <span class="dv">8</span> <span class="co"># number of important variables with nonzeros coefficients</span></span>
<span id="cb13-8"><a href="#cb13-8" tabindex="-1"></a>nonzero <span class="ot">=</span> <span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span>p, k) <span class="co"># randomly select 8 important variables</span></span>
<span id="cb13-9"><a href="#cb13-9" tabindex="-1"></a></span>
<span id="cb13-10"><a href="#cb13-10" tabindex="-1"></a>beta <span class="ot">=</span> <span class="fu">as.numeric</span>(<span class="dv">1</span><span class="sc">:</span>p <span class="sc">%in%</span> nonzero) <span class="co"># assign non-zero coefficients to the important variables</span></span>
<span id="cb13-11"><a href="#cb13-11" tabindex="-1"></a>y <span class="ot">=</span> X <span class="sc">%*%</span> beta </span></code></pre></div>
<div id="step-1-knockoff-variable-generation" class="section level3">
<h3>Step 1: Knockoff Variable Generation</h3>
<p>First, we generate the knockoff variables using the function. The
function requires the predictor matrix as input and returns a matrix of
knockoff variables.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="co"># generate the knockoff variables</span></span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a>plsko_default <span class="ot">=</span> <span class="fu">plsko</span>(X) </span></code></pre></div>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
